### ISSUE TITLE
Publish the measurement harnesses: the repo cannot currently reproduce its own benchmark records

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#379]
## What

`benchmarks/gate-h100-validation.md` closes with a "Reproducing" section that lists roughly twenty harness scripts — and then says they **live in the session scratchpad, not in the repo** (L4908–4939):

```
bitexact.py  graddiff.py  determinism.py  repeat_backward.py  layercount.py
depth_vs_bytes.py  ckpt_hypothesis.py  mechanism.py  mechanism_cost.py  pincost.py
prep_convergence.py  runbench.sh  variance.sh  runbench8.sh  numerics.py  numerics2.py
forcecheck.py  fixtureshape.py  fixture_window.py  cpu_mode_probe.py  variant2_gate.py
bnb_repro.py  issue328_min.py  issue328_probe.py  issue328_control.py  measure327.py
```

**Nothing in the repository reproduces the record.** The same is true of the four earlier gate records in `benchmarks/`.

## Why this is the largest credibility gap in the whole harvest

The project's entire public position is *check it yourself*. `README.md` points at `benchmarks/` as the measurement record behind the preprint (DOI 10.5281/zenodo.21771064). A reader who wants to verify a number — the 256/256-vs-8/256 repair gate, the 6.56x pinning cost, the 163.8–171.5 MiB threshold bracket — currently cannot, because the code that produced it does not exist publicly.

Every one of those numbers is reproducible in principle: `bnb_repro.py` reproduces the NF4 defect **in about a minute on a synthetic 7.7 GB model with no download**.

## Proposal

`benchmarks/harness/` with:

- the scripts as run, not cleaned up,
- a README mapping **script → STEP → the claim it produced**,
- the hardware and stack each was run against,
- and the deliberately-vacuous ones **kept with their vacuity documented** — `numerics.py` measured at M=2048, above `_gemm_4bit_custom_max_m`, so both arms ran the same fallback and it compared variant 2 against variant 2. That is exactly the kind of artefact the records already preserve in prose; deleting it from the code while keeping it in the text would be inconsistent.

All harnesses `import soup_cli` unmodified, which is worth stating explicitly: every measurement above the record's measurement/fix boundary was taken against untouched Soup.

## Acceptance criteria

- [ ] Scripts published under `benchmarks/harness/`, indexed in `benchmarks/README.md`
- [ ] At least one — `bnb_repro.py` — documented as runnable end to end on a single card with no download, since it is the cheapest independent check of the session's central finding
- [ ] Scripts that require multiple GPUs or a specific stack say so at the top
- [ ] No secrets, no absolute paths from the session box

**Comment by MakazhanAlpamys** [comment]
## Acceptance item added, from #369 / #384

#369 is the concrete cost of this issue, so the fix for it should not be allowed to evaporate when it closes.

The comparison that reported `grads exact 0/0` — the #331 repair gate at 32B reading an **empty** streamed-vs-resident intersection as a pass (`benchmarks/gate-h100-validation.md` L2627-2643) — lives in the harness behind that record, and the harness is what this issue is about publishing. #384 ships `assert_canonical_parameters_intersect()`, which raises on an empty intersection instead of returning success on nothing compared.

But a helper nobody is obliged to call does not prevent anything. As things stand #369's second criterion is satisfied by **availability**, not by **enforcement**: the next harness written against a streamed model can compare raw `named_parameters()` and produce the same false green, and the same control-arm accident is what would have to catch it again.

So, added to this issue's scope:

- [ ] The published harness compares streamed and resident parameters through `layer_stream_runtime.assert_canonical_parameters_intersect()` (or `canonical_named_parameters()` on both sides), never raw `named_parameters()`
- [ ] Its own test suite includes the negative: a comparison over an empty intersection fails the harness rather than reporting a pass

The second one is the load-bearing half. The reason this defect survived to a 32B run was not that anybody reasoned wrongly about names — it was that `n_exact == n_compared` is true at `0 == 0` and nothing in the harness treated "nothing was compared" as a distinct outcome from "everything matched".

**Comment by Samearth17** [comment]
I would like to take #379.

I can start by expanding `benchmarks/harness/` with the remaining reproducible measurement harnesses, while preserving the original measurement methodology. I will also update the harness README so each script is mapped to the benchmark record/claim it reproduces and document the required hardware and software.

I will make sure the streamed-vs-resident comparisons use the canonical parameter helpers from #384 and fail on an empty parameter intersection, as required by the updated acceptance criteria.

I will keep the work incremental and focused on reproducibility rather than changing the underlying benchmark methodology.

**Comment by MakazhanAlpamys** [comment]
Claim noted, thank you. Please keep the first draft reproducibility-only: publish the harnesses and their index without rewriting the measurement methodology. Include bnb_repro.py as the runnable no-download path, state each hardware/software requirement at the script boundary, and add the empty-intersection negative test wherever streamed and resident parameters are compared. A draft PR is welcome once that first slice is ready.

**Comment by Samearth17** [comment]
Draft PR #589 is up with the first reproducibility-only slice for #379.

The first slice publishes `benchmarks/harness/bnb_repro.py` as the standalone upstream bitsandbytes #2034 reproducer and updates `benchmarks/README.md` to index the harness and document its hardware/software requirements.

The harness preserves the upstream reproduction protocol: NF4 with recycled buffers, NF4 with private buffers as the reference, and bf16 with recycled buffers as the control. No model download is required, and the script explicitly requires CUDA.

I also kept the change limited to reproducibility artifacts; no benchmark methodology or underlying Soup implementation was changed.

The full Soup test suite passes locally (`18,901 passed, 283 skipped, 5 deselected`), Ruff passes, and the harness correctly reports that CUDA is required on my Mac rather than claiming an end to end reproduction there.

**Comment by MakazhanAlpamys** [comment]
First slice merged: PR #589 (`25d17c0`) by @Samearth17 publishes `benchmarks/harness/bnb_repro.py`, the standalone upstream bitsandbytes #2034 NF4 reproducer, and indexes it in `benchmarks/README.md`. Methodology and the existing `issue331_qlora_scope.py` row are untouched, as asked.

**This issue stays open** — the harness index now has two entries out of roughly twenty, so the repo still cannot reproduce most of its own records.

Verified on real CUDA before merge (RTX 3050 Laptop, torch 2.5.1+cu121, bitsandbytes 0.49.2), since the contributor has no GPU:

```
normal run                        exit 0, NF4 mismatch 6/8 layers, bf16 control clean 8/8
--bypass-pool                     exit 1, negative control detected
pool.acquire() cut from source    exit 1
same cut on pre-fix 373c007e      exit 0
```

The last two lines are the reason the review round happened: a harness that exits `0` whether or not the defect is present is not evidence of anything. It now fails in both directions — an explicit `--bypass-pool` control, and an ordinary invocation that requires the mismatch.

One finding for whoever takes the next slice: the reproduction holds on torch 2.5.1+cu121 / bnb 0.49.2, well below the torch 2.13.0+cu130 / bnb 0.50.0 environment the record was taken on, so the mechanism is not specific to the newer stack. Also, layers 6 and 7 read exactly `0.000000e+00` on a 6-slot pool — plausibly because the last two layers still hold live buffers, but that is unconfirmed and should not be asserted without measuring it.

Remaining harnesses from this issue are unclaimed. Anyone picking one up: say which here first, keep it reproducibility-only, and make the default invocation fail when the thing it demonstrates is absent.

**Comment by MakazhanAlpamys** [comment]
**First harness published** — #597 by @Samearth17, merged as `4021b9f`. This issue **stays open**: it asks for *the harnesses*, plural, and one of roughly twenty is now in the tree.

## What landed

`benchmarks/harness/bitexact.py` — shard -> stream -> compare logits, LoRA gradients and the loss curve against a resident reference at matching numerics, with a row in the `## Harnesses` table naming its software versions.

Run on real CUDA before merge, both directions:

```
                                    positive          --skip-lora-sync
adapter_tensors_copied              120               0
max_abs_logit_diff                  0.0               3.687500e+00
exit                                0                 1
```

The negative direction is the point. A harness whose only output is "these two agree to 0.0" has to be **able to report disagreement**, or it is a script that prints a reassuring line. `--skip-lora-sync` is a built-in control for exactly that, and `make_non_vacuous_lora` closes the other axis so "gradients exact" cannot mean "both sides are zero".

## What is still open here

Every other harness from the v0.72.x gate sessions. Named ones people have already run into:

- the **STEP 20 FSDP per-rank sharding probe** — #618 is in review for it, and note the probe as recorded is not merely unrun but *wrong* (see the correction on #373: it hits a `torch.compile` wrapper, not a bare FSDP module)
- the throughput / GEMM-ceiling probes behind the streaming forecast panel (see #617)
- the VRAM-fit sweep — partially covered by `issue395_second_stack_vram.py` in #610, in review

## Coordination note, because this issue is plural and has already caused one near-collision

Nobody owns every future script in `benchmarks/harness/`. What a claim on this issue means is **the specific harness you are publishing**, not the directory. @Srinivasan8888 raised exactly this on #373 before writing anything, which is the right instinct and is why there was no collision.

**Say which harness you are taking when you claim**, and the only real contact point between two contributors is one row in the `benchmarks/README.md` table.

**Comment by MakazhanAlpamys** [comment]
@Samearth17 — status check. #589 and #597 published two of roughly twenty harnesses; nothing has moved since 09-01.

Two things that make this smaller than when you claimed it:

- The publishing pattern is now established — `benchmarks/README.md` has an index table, and #1014 (merged today) is a clean recent example of a harness landing with its row, its scope and its stated limits.
- Several of the remaining harnesses are for measurements this project has since taken, so the record exists and it is the publishing step that is missing, not the work.

Still yours, or would you rather it be split so several people can take a harness each? Either answer is useful; silence is the only one that is not. See also my note on #596 — you are holding four of these at once and I would rather help you finish one than keep all four reserved.

**Comment by Samearth17** [comment]
I’m going to release #379 for now so the remaining harnesses can be distributed among contributors.

I’ve had some health issues recently and haven’t been able to move as quickly as I expected. There are roughly 20 harnesses in scope, with the first 2 already published. I can still take the following 8 remaining slices and work through them incrementally:

* `mechanism.py`
* `mechanism_cost.py`
* `pincost.py`
* `graddiff.py`
* `determinism.py`
* `depth_vs_bytes.py`
* `repeat_backward.py`
* `layercount.py`

The other remaining harnesses can be split among other contributors, with each person claiming the specific harness they plan to publish.

For the slices I take, I’ll keep the work reproducibility-only, preserve the existing measurement methodology, and follow the established `benchmarks/harness/` + `benchmarks/README.md` publishing pattern.

**Comment by taljeon** [comment]
I can take the unreserved `variant2_gate.py` slice of #379. This does not overlap the eight harnesses @Samearth17 retained (`mechanism*`, `pincost`, `graddiff`, `determinism`, `depth_vs_bytes`, `repeat_backward`, and `layercount`).

Plan: publish the STEP 14 control-versus-repair harness from `gate-h100-validation.md`, keeping the control and repaired arms in one process against one resident reference. The harness will use `assert_canonical_parameters_intersect()` / canonical parameter names and treat an empty gradient intersection as a hard failure, with a load-bearing negative test. I’ll add the harness index row and focused CPU tests without changing Soup's training implementation or the recorded methodology.

This host has no CUDA GPU, so I will not claim a new 32B/72B measurement. I will run the pure/CLI/static tests and verify the intentional no-CUDA path; the PR will state that real CUDA execution remains external validation before merge.

AI disclosure: I’m using OpenAI Codex (`gpt-5.6-sol`) to assist with source inspection, implementation, and verification. I’ll review the exact diff, use my GitHub noreply identity, and report only checks actually run.

**Comment by MakazhanAlpamys** [comment]
Acknowledging the split, so that the claims are written down in one place:
- **@Samearth17** keeps `mechanism.py`, `mechanism_cost.py`, `pincost.py`, `graddiff.py`, `determinism.py`, `depth_vs_bytes.py`, `repeat_backward.py` and `layercount`. Thank you for releasing the rest explicitly rather than holding it; take care of yourself first.
- **@taljeon** has `variant2_gate.py`, in review as #1085.

Every other harness in the tracker is open. Please claim a slice here by file name before starting, as @taljeon did.

**Comment by Amix29** [comment]
I’ll take the unreserved `fixture_window_cpu.py` slice.

Scope is reproducibility-only: publish the STEP 13 CPU fixture-window probe for the shipped 64x64 / 256x64 NF4 fixture shapes across the recorded M sweep, compare bitsandbytes against the explicit dequantize+linear variant, and include a path/control signal so equality cannot mean both arms accidentally took the same unobserved route.

I’ll index it in `benchmarks/README.md` and add dependency-light tests for the shape/grid/output contract. This Mac can execute the CPU harness end to end, so I’ll report only measurements actually reproduced here.

**Comment by Amix29** [comment]
The claimed fixture_window_cpu.py slice is published as draft #1144. On this Apple-Silicon host it runs all 14 recorded shape/M rows, but AVX512-BF16 is absent so packed_inference_rows=0; all pairwise differences are 0.0 here. The harness records that path fact and --require-packed-inference exits 2, so this result cannot be misrepresented as a reproduction of the historical packed-CPU divergence. 4 focused tests pass and the harness is indexed in benchmarks/README.md.

**Comment by Amix29** [comment]
Update on my claimed slice: PR #1144 now covers both the recovered `fixture_window_cpu.py` and the `cpu_mode_probe.py` mechanism question. The final harness computes a same-run emulated-bf16 reference per row, pins the 2% envelope inputs/boundaries, records CPU-kernel attribution, and conditionally exercises the real AVX512 success path. Mac validation: 30 passed / 1 AVX512 skip; 12-row survey gives worst emulated bf16 rel 0.479%, so the 2% envelope is 4.17x that same-run reference.

**Comment by MakazhanAlpamys** [comment]
Claims, back in one place (the last summary was on 09-19):

- @Samearth17: `mechanism_cost.py` merged in #1121, thank you. `mechanism.py` (#1119), `graddiff.py` (#1170) and `pincost.py` (#1172) are in review with changes requested. `determinism.py`, `depth_vs_bytes.py`, `repeat_backward.py` and `layercount` stay reserved to you.
- @taljeon: `variant2_gate.py`, draft #1085.
- @Amix29: `fixture_window_cpu.py` and, as of today's update, `cpu_mode_probe.py`, both in draft #1144.

Every other harness in the tracker is open. Claim one here by file name before starting.
